src/Extensions/Diagnostics.HealthChecks/Internal/ServiceFabricHealthCheckPublisher.cs

```diff
@@ -34,61 +34,99 @@ internal class ServiceFabricHealthCheckPublisher : IHealthCheckPublisher
 		public Task PublishAsync(HealthReport report, CancellationToken cancellationToken)
 		{
 			IServicePartition? partition = m_partitionAccessor.Value;
-
 			if (partition == null)
 			{
-				m_logger.LogWarning(Tag.Create(), "Publisher run before partition provided");
+				m_logger.LogWarning(Tag.Create(), "Publisher run before partition provided.");
 
 				return Task.CompletedTask;
 			}
 
-			partition.ReportPartitionHealth(new SfHealthInformation(SourceId, "Summary", ConvertStatus(report.Status)));
-
-			foreach(KeyValuePair<string, HealthReportEntry> entryPair in report.Entries)
+			// Avoiding repeated pattern matching for each report entry.
+			Action<SfHealthInformation> reportHealth = partition switch
 			{
-				HealthReportEntry entry = entryPair.Value;
-				SfHealthInformation sfHealthInformation = new SfHealthInformation(SourceId, entryPair.Key, ConvertStatus(entry.Status))
-				{
-					Description = CreateDescription(entry)
-				};
+				IStatefulServicePartition statefulPartition => statefulPartition.ReportReplicaHealth,
+				IStatelessServicePartition statelessPartition => statelessPartition.ReportInstanceHealth,
+				_ => (_) => { }, // NOOP, not going to happen, but we need a default case for the switch.
+			};
 
-				try
-				{
-					partition.ReportPartitionHealth(sfHealthInformation);
-				}
-				catch (Exception exception)
+			try
+			{
+				foreach (KeyValuePair<string, HealthReportEntry> entryPair in report.Entries)
 				{
-					m_logger.LogError(Tag.Create(), exception, "Failed to report partition health");
+					cancellationToken.ThrowIfCancellationRequested();
+					reportHealth(BuildSfHealthInformation(entryPair.Key, entryPair.Value));
 				}
+
+				cancellationToken.ThrowIfCancellationRequested();
+				reportHealth(BuildSfHealthInformation(report));
+			}
+			catch (FabricObjectClosedException)
+			{
+				// Ignore, the service instance is closing.
 			}
 
 			return Task.CompletedTask;
 		}
 
-		private string CreateDescription(HealthReportEntry entry)
+		private SfHealthInformation BuildSfHealthInformation(string healthCheckName, HealthReportEntry reportEntry)
 		{
-			// Healthy reports won't be displayed and most of the checks will be healthy,
-			// so we are creating a detailed description only for failed checks and avoiding allocations for healthy
-			if (entry.Status == HealthStatus.Healthy)
+			StringBuilder descriptionBuilder = new StringBuilder();
+			if (!string.IsNullOrWhiteSpace(reportEntry.Description))
 			{
-				return entry.Description;
+				descriptionBuilder.AppendLine($"Description: {reportEntry.Description}");
```

Comment:
I think calling `Append` and `AppendLine` would be for performant then using `$`

Reply:
Maybe, but much less readable. How important is it to save a few nanoseconds every few minutes?

Reply:
It's not about nanoseconds, it's about allocating fewer objects, since allocating more will increase the rates of GC calls that will stop all of your applications to collect objects, I think infrastructure should reduce it's contributions to this process as much as posible

Reply:
Changed to use .Append...